builder/amazon/common/ami_config.go

```diff
@@ -170,26 +170,33 @@ func (c *AMIConfig) Prepare(accessConfig *AccessConfig, ctx *interpolate.Context
 		}
 	}
 
-	var kmsKeys []string
+	kmsKeys := make([]string, 0)
 	if len(c.AMIKmsKeyId) > 0 {
 		kmsKeys = append(kmsKeys, c.AMIKmsKeyId)
 	}
 	if len(c.AMIRegionKMSKeyIDs) > 0 {
 		for _, kmsKey := range c.AMIRegionKMSKeyIDs {
-			if len(kmsKey) == 0 {
-				kmsKeys = append(kmsKeys, c.AMIKmsKeyId)
+			if len(kmsKey) > 0 {
+				kmsKeys = append(kmsKeys, kmsKey)
 			}
 		}
 	}
+
+	if len(kmsKeys) > 0 && !c.AMIEncryptBootVolume.True() {
+		errs = append(errs, fmt.Errorf("If you have set either "+
+			"region_kms_key_ids or kms_key_id, encrypt_boot must also be true."))
```

Comment:
Taking a step back here; would it be a good thing to default `AMIEncryptBootVolume` to `true` if it is nil ?

Reply:
I did think about it. But I think I'd rather force the user to be explicit; this behavior is so complex and has been such a source of bugs/unexpected behaviors I want users to know what they're getting in to. Like, if we silently default it to True, and they want to leave some kmskeys blank, then we'll use the default key for region copy even if they started from an unencrypted source ami, which is against the documentation for what we say we'll do if the amiencryptbootvolume is nil. The better behavior is to force people to understand that if they encrypt for one region they encrypt for all regions.

Reply:
Gotcha ! 🙂